File: src/main/java/com/naminhyeok/fantazzk/room/RoomPlayer.java
package com.naminhyeok.fantazzk.room;

import jakarta.persistence.Access;
import jakarta.persistence.AccessType;
import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Embeddable;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import lombok.Getter;

@Getter
@Access(AccessType.FIELD)
@Embeddable
class RoomPlayer {
    @Column(name = "room_player_id")
    @Convert(converter = RoomPlayerId.JpaConverter.class)
    private RoomPlayerId id;
    @Column(name = "name")
    private String name;
    @Column(name = "position")
    private String position;
    @Column(name = "display_order")
    private int displayOrder;
    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    private PlayerStatus status;

    RoomPlayer() {
    }

    RoomPlayer(RoomPlayerId id, String name, String position, int displayOrder) {
        this.id = id;
        this.name = name;
        this.position = position;
        this.displayOrder = displayOrder;
        this.status = PlayerStatus.AVAILABLE;
    }

    RoomPlayer(RoomPlayerId id, String name, String position, int displayOrder, PlayerStatus status) {
        this.id = id;
        this.name = name;
        this.position = position;
        this.displayOrder = displayOrder;
        this.status = status;
    }

    public RoomPlayerId getId() {
        return id;
    }

    void assign() {
        this.status = PlayerStatus.ASSIGNED;
    }

    void moveToBack(int nextOrder) {
        this.displayOrder = nextOrder;
    }

    void sync(int displayOrder, PlayerStatus status) {
        this.displayOrder = displayOrder;
        this.status = status;
    }
}

File: src/main/kotlin/com/naminhyeok/fantazzk/room/RoomPlayer.kt
package com.naminhyeok.fantazzk.room

import java.time.Instant

data class RoomPlayer(
    val roomPlayerId: Long = 0L,
    val roomId: Long,
    val name: String,
    val status: PlayerStatus = PlayerStatus.AVAILABLE,
    val displayOrder: Int,
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now(),
) {
    fun assign(): RoomPlayer {
        check(status == PlayerStatus.AVAILABLE) { "선수를 배정할 수 없습니다" }
        return copy(status = PlayerStatus.ASSIGNED)
    }

    fun moveToBack(displayOrder: Int): RoomPlayer {
        check(status == PlayerStatus.AVAILABLE) { "선수를 뒤로 보낼 수 없습니다" }
        require(displayOrder >= 0) { "순서는 0 이상이어야 합니다" }
        require(displayOrder > this.displayOrder) { "현재 순서보다 뒤로만 이동할 수 있습니다" }
        return copy(displayOrder = displayOrder)
    }

}

fun RoomPlayer.isAvailable(): Boolean = status == PlayerStatus.AVAILABLE
